File: src/audio_stegano_desktop/src/AudioStegano.Desktop/Services/PlaybackHub.cs
using AudioStegano.Core.Audio;

namespace AudioStegano.Desktop.Services;

/// <summary>
/// One isolated engine per UI play surface — method-based exclusivity.
/// </summary>
public enum PlaybackSessionId
{
    EmbedCover,
    EmbedStego,
    EmbedPayloadOriginal,
    EmbedPayloadRecovered,
    ExtractCover,
    ExtractPayload,
}

/// <summary>
/// Each session owns its <see cref="AudioPlaybackService"/>; Play pauses others first.
/// </summary>
public sealed class PlaybackHub
{
    public static PlaybackHub Instance { get; } = new();

    private readonly Dictionary<PlaybackSessionId, AudioPlaybackService> _engines = new();

    public static readonly PlaybackSessionId[] EmbedSessions =
    [
        PlaybackSessionId.EmbedCover,
        PlaybackSessionId.EmbedStego,
        PlaybackSessionId.EmbedPayloadOriginal,
        PlaybackSessionId.EmbedPayloadRecovered,
    ];

    public static readonly PlaybackSessionId[] ExtractSessions =
    [
        PlaybackSessionId.ExtractCover,
        PlaybackSessionId.ExtractPayload,
    ];

    public static readonly PlaybackSessionId[] AbSessions =
    [
        PlaybackSessionId.EmbedCover,
        PlaybackSessionId.EmbedStego,
    ];

    private PlaybackHub()
    {
    }

    public AudioPlaybackService Engine(PlaybackSessionId id)
    {
        if (!_engines.TryGetValue(id, out var engine))
        {
            engine = new AudioPlaybackService();
            _engines[id] = engine;
        }
        return engine;
    }

    public bool IsPlaying(PlaybackSessionId id) =>
        _engines.TryGetValue(id, out var e) && e.IsPlaying;

    public bool HasSource(PlaybackSessionId id) =>
        _engines.TryGetValue(id, out var e) && e.HasSource;

    public bool IsPaused(PlaybackSessionId id) =>
        _engines.TryGetValue(id, out var e) && e.IsPaused;

    /// <summary>Play [wav] on [id]; pause every other session first.</summary>
    public void Play(PlaybackSessionId id, WavFile wav)
    {
        PauseOthers(id);
        Engine(id).Play(wav);
    }

    /// <summary>
    /// Toggle: playing → pause; paused with source → resume; else load+play.
    /// </summary>
    public void PlayOrToggle(PlaybackSessionId id, WavFile wav)
    {
        var e = Engine(id);
        if (e.IsPlaying)
        {
            e.Pause();
            return;
        }
        if (e.HasSource && e.IsPaused)
        {
            PauseOthers(id);
            e.Resume();
            return;
        }
        Play(id, wav);
    }

    /// <summary>Do not restart when already playing this session.</summary>
    public void PlayIfNotPlaying(PlaybackSessionId id, WavFile wav)
    {
        var e = Engine(id);
        if (e.IsPlaying) return;
        if (e.HasSource && e.IsPaused)
        {
            PauseOthers(id);
            e.Resume();
            return;
        }
        Play(id, wav);
    }

    public void Pause(PlaybackSessionId id)
    {
        if (_engines.TryGetValue(id, out var e))
            e.Pause();
    }

    public void Stop(PlaybackSessionId id)
    {
        if (_engines.TryGetValue(id, out var e))
            e.Stop();
    }

    public void PauseOthers(PlaybackSessionId keep)
    {
        foreach (var (sessionId, engine) in _engines)
        {
            if (sessionId == keep) continue;
            if (engine.IsPlaying)
                engine.Pause();
        }
    }

    public void StopSessions(IEnumerable<PlaybackSessionId> ids)
    {
        foreach (var id in ids)
            Stop(id);
    }

    public void StopAll() => StopSessions(Enum.GetValues<PlaybackSessionId>());

    public void PauseAbSessions()
    {
        foreach (var id in AbSessions)
            Pause(id);
    }

    public void StopAbSessions()
    {
        foreach (var id in AbSessions)
            Stop(id);
    }
}

File: src/audio_stegano_app/lib/core/audio/playback_hub.dart
import 'dart:async';

import 'package:just_audio/just_audio.dart';

import 'audio_player.dart';
import 'wav_io.dart';

/// One isolated engine per UI play surface — no shared mutable play source.
enum PlaybackSessionId {
  embedCover,
  embedStego,
  embedPayloadOriginal,
  embedPayloadRecovered,
  extractCover,
  extractPayload,
}

/// Method-based exclusive playback: each session owns its [AudioPlayerService];
/// [play] / [playOrToggle] pause every other session first.
final class PlaybackHub {
  PlaybackHub._();
  static final PlaybackHub instance = PlaybackHub._();

  final Map<PlaybackSessionId, AudioPlayerService> _engines = {};

  static const embedSessions = <PlaybackSessionId>[
    PlaybackSessionId.embedCover,
    PlaybackSessionId.embedStego,
    PlaybackSessionId.embedPayloadOriginal,
    PlaybackSessionId.embedPayloadRecovered,
  ];

  static const extractSessions = <PlaybackSessionId>[
    PlaybackSessionId.extractCover,
    PlaybackSessionId.extractPayload,
  ];

  static const abSessions = <PlaybackSessionId>[
    PlaybackSessionId.embedCover,
    PlaybackSessionId.embedStego,
  ];

  AudioPlayerService engine(PlaybackSessionId id) =>
      _engines.putIfAbsent(id, AudioPlayerService.new);

  bool isPlaying(PlaybackSessionId id) => _engines[id]?.isPlaying ?? false;

  bool hasSource(PlaybackSessionId id) => _engines[id]?.hasSource ?? false;

  PlaybackSessionId? get activePlaying {
    for (final id in PlaybackSessionId.values) {
      if (isPlaying(id)) return id;
    }
    return null;
  }

  /// Start [wav] on [id]; pauses all other sessions (no cross-button swap).
  Future<void> play(PlaybackSessionId id, WavFile wav) async {
    await pauseOthers(id);
    await engine(id).playWav(wav);
  }

  /// If [id] is playing → pause. If paused with source → resume (after pausing
  /// others). Otherwise load and play [wav].
  Future<void> playOrToggle(PlaybackSessionId id, WavFile wav) async {
    final e = engine(id);
    if (e.isPlaying) {
      await e.pause();
      return;
    }
    if (e.hasSource) {
      await pauseOthers(id);
      await e.resume();
      return;
    }
    await play(id, wav);
  }

  /// Play without restarting when already playing the same session.
  Future<void> playIfNotPlaying(PlaybackSessionId id, WavFile wav) async {
    final e = engine(id);
    if (e.isPlaying) return;
    if (e.hasSource) {
      await pauseOthers(id);
      await e.resume();
      return;
    }
    await play(id, wav);
  }

  Future<void> pause(PlaybackSessionId id) async {
    await _engines[id]?.pause();
  }

  Future<void> stop(PlaybackSessionId id) async {
    await _engines[id]?.stop();
  }

  Future<void> pauseOthers(PlaybackSessionId keep) async {
    final futures = <Future<void>>[];
    for (final entry in _engines.entries) {
      if (entry.key == keep) continue;
      if (entry.value.isPlaying) {
        futures.add(entry.value.pause());
      }
    }
    if (futures.isNotEmpty) await Future.wait(futures);
  }

  Future<void> pauseSessions(Iterable<PlaybackSessionId> ids) async {
    await Future.wait(ids.map((id) async {
      await _engines[id]?.pause();
    }));
  }

  Future<void> stopSessions(Iterable<PlaybackSessionId> ids) async {
    await Future.wait(ids.map((id) async {
      await _engines[id]?.stop();
    }));
  }

  Future<void> stopAll() => stopSessions(PlaybackSessionId.values);

  StreamSubscription<PlayerState> listenState(
    PlaybackSessionId id,
    void Function(PlayerState state) onData,
  ) {
    return engine(id).stateStream.listen(onData);
  }

  StreamSubscription<List<double>> listenSpectrum(
    PlaybackSessionId id,
    void Function(List<double> bands) onData,
  ) {
    return engine(id).spectrumStream.listen(onData);
  }
}
